feat(cli): configurable certgen images

akunszt opened the pull request:
**What this PR does / why we need it**:

Add two new command line options to the manager.

```
  -webhook-certificate-config-base-image string
    	The base image for the certgen Jobs. (default "registry.k8s.io/ingress-nginx/kube-webhook-certgen:v1.3.0")
  -webhook-certificate-config-shell-image string
    	The shell image for the certgen Jobs. (default "busybox")
```

Those are optional. If you omit them the defaults will be used which are the previous hard coded values.

By using these command line arguments the user can deploy kong-gateway-operator to a cluster where external Docker repositories are not reachable or prohibited to use them.

**Which issue this PR fixes**

Fixes #515.

**Special notes for your reviewer**:

I don't think this is a significant change and as this is my first PR I am not sure if I have to touch CHANGELOG.md or not.

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [x] the `CHANGELOG.md` release notes have been updated to reflect significant changes

akunszt committed:
feat(cli): configurable certgen images

Add two new command line options to the manager.

  -webhook-certificate-config-base-image string
    	The base image for the certgen Jobs. (default "registry.k8s.io/ingress-nginx/kube-webhook-certgen:v1.3.0")
  -webhook-certificate-config-shell-image string
    	The shell image for the certgen Jobs. (default "busybox")

Those are optional. If you omit them the defaults will be used which are
the previous hard coded values.

Signed-off-by: Arpad Kunszt <akunszt@hiya.com>

pmalek commented on modules/cli/cli.go:
```diff
@@ -167,6 +171,9 @@ func (c *CLI) Parse(arguments []string) manager.Config {
 	}
 
+	webhookCertificateConfigBaseImage := c.deferFlagValues.WebhookCertificateConfigBaseImage
+	webhookCertificateConfigShellImage := c.deferFlagValues.WebhookCertificateConfigShellImage
```
You shouldn't really need this here. Those 2 flags can simply be used as is.

This func is for flags that require more logic than just binding to a variable.

akunszt replied:
Thanks for the quick response. I modified the code as you requested.

pmalek reviewed (commented):
The proposal looks reasonable. I made a comment about those flags not requiring to be set in the `Parse` func.

As for the changelog: yes, that would most certainly be a nice change to place it in there.

Thanks!

akunszt committed:
feat(cli): cleanup code per review

Removed unnecessary variable assignments.

Signed-off-by: Arpad Kunszt <akunszt@hiya.com>

akunszt committed:
feat(cli): add CHANGELOG entry

Signed-off-by: Arpad Kunszt <akunszt@hiya.com>

akunszt committed:
feat(cli): unit tests handle new arguments

Signed-off-by: Arpad Kunszt <akunszt@hiya.com>

czeslavo commented on modules/cli/cli.go:
```diff
@@ -52,6 +52,8 @@ func New(m metadata.Info) *CLI {
 	// webhook and validation options
 	flagSet.BoolVar(&deferCfg.ValidatingWebhookEnabled, "enable-validating-webhook", true, "Enable the validating webhook.")
+	flagSet.StringVar(&deferCfg.WebhookCertificateConfigBaseImage, "webhook-certificate-config-base-image", consts.WebhookCertificateConfigBaseImage, "The base image for the certgen Jobs.")
+	flagSet.StringVar(&deferCfg.WebhookCertificateConfigShellImage, "webhook-certificate-config-shell-image", "busybox", "The shell image for the certgen Jobs.")
```
Please consider extracting `busybox` to a constant, similarly to `consts.WebhookCertificateConfigBaseImage`, and using it in UTs.

pmalek commented on modules/cli/cli.go:
```diff
@@ -88,10 +90,12 @@ type CLI struct {
+	ValidatingWebhookEnabled           bool
+	Version                            bool
+	WebhookCertificateConfigBaseImage  string
+	WebhookCertificateConfigShellImage string
```
This shouldn't be necessary as previously mentioned. Just add it to the `Config` struct: https://github.com/Kong/gateway-operator/blob/fa31122c75a4ff06710f7f71ce666c771bfc34f4/modules/manager/run.go#L61

pmalek commented on modules/cli/cli.go:
```diff
@@ -52,6 +52,8 @@ func New(m metadata.Info) *CLI {
 	// webhook and validation options
 	flagSet.BoolVar(&deferCfg.ValidatingWebhookEnabled, "enable-validating-webhook", true, "Enable the validating webhook.")
+	flagSet.StringVar(&deferCfg.WebhookCertificateConfigBaseImage, "webhook-certificate-config-base-image", consts.WebhookCertificateConfigBaseImage, "The base image for the certgen Jobs.")
+	flagSet.StringVar(&deferCfg.WebhookCertificateConfigShellImage, "webhook-certificate-config-shell-image", "busybox", "The shell image for the certgen Jobs.")
```
Those 2 flags should be bound to the `Config` struct instead of the defer config as there's nothing to evaluate here.

akunszt replied:
Got it. I just misunderstood the `flagsForFurtherEvaluation`. I thought that those are the flags which are still evaluated to keep, modify or drop by the team. But it is clear that further evaluation means that flags which needs some extra logic and cannot be assigned directly to the configuration.

I will create unit tests and then push the changes.

pmalek reviewed (changes requested):
Looking good.

Left a couple of comments to move the flags to `Config`.

Please add a handful of UTs for those flags in https://github.com/Kong/gateway-operator/blob/a0614c5216d8ed3f83d6186365fa2b0d805443ea/modules/cli/cli_test.go

akunszt committed:
feat(cli): move new configuration into Config

Also created a constant for the shell image, so it is no more a hard
coded string hidden in the code.

Signed-off-by: Arpad Kunszt <akunszt@hiya.com>

akunszt committed:
feat(cli): added UT for the new arguments

The new unit test tests only if the command line arguments are set in
the configuration. The other cases, the default configuration, the
environmental variable handling are already tested in previous cases.

The test does not cover if the values from the configuration are
actually used but at the moment there are no tests for that part of the
code at all.

Signed-off-by: Arpad Kunszt <akunszt@hiya.com>

pmalek commented:
Thanks @akunszt 🙇
